Automattic/wp-calypso
# Extract the search component into `@automattic/search`

#### Rationale

We need to extract this component so that it can be used inside of the language-picker component. I've also moved it outside of `@automattic/components` for the same reasons laid out here: https://github.com/Automattic/wp-calypso/pull/46020#issuecomment-707869250

#### Changes proposed in this Pull Request

* Extract the search component from Calypso into a separate package
* Removes analytics from the component. Consumers will need to implement their own using the appropriate hooks.

This is more or less a 1-to-1 reimplementation of the existing component. There are some differences, however:
* Updated ref usage
* Using WordPress components rather than Calypso components and icons
* Using TypeScript and fixing some potential coding errors
* Add a wrapping noop `form` to cause iOS to display a "Search" button rather than "return"

<img height="500" alt="iOS displaying a Search button instead of Return on the soft keyboard" src="https://user-images.githubusercontent.com/24264157/96284364-545db000-0f92-11eb-92a0-1351d1425d9d.jpeg"></img>


#### Testing instructions



* Run the new storybook command `yarn run search:storybook:start` and ensure the examples work as expected and cover the different use cases

Note: I didn't port over the existing tests because they aren't really that useful of tests.

#### Questions

1. The way this component is written it can go from controlled to uncontrolled (if you start to pass in value later in the component lifecycle. We could vastly simplify this component if we turned it into a fully uncontrolled component. For example, if we did this, we could remove the derived state logic and the "responsiveness" of the component to certain props changing (like do we really need to be able to handle changes in `delaySearch`? That seems like a waste and a supremely unlikely edge case usage). Shall we move forward with turning this component into an uncontrolled component (i.e., by not responding to changes in "static" props and not allowing `value` to be passed in?)

Fixes #
